src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties': {
+        Object.assign(stateCopy, action.extraData);
+        break;
+      }
+
+      case 'removeProperties': {
+        action.keysToRemove.forEach(key => delete stateCopy[key]);
+        break;
+      }
+
+      case 'clear': {
+        Object.keys(stateCopy).forEach(key => delete stateCopy[key]);
```

Comment:
[DRY](https://www.plutora.com/blog/understanding-the-dry-dont-repeat-yourself-principle)